// lit-html/development/directive.js
/**
 * @license
 * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
import { _$private, } from './lit-html';
const resolveDirective = _$private._resolveDirective;
export const PartType = {
    ATTRIBUTE: 1,
    CHILD: 2,
    PROPERTY: 3,
    BOOLEAN_ATTRIBUTE: 4,
    EVENT: 5,
    ELEMENT: 6,
};
/**
 * Creates a user-facing directive function from a Directive class. This
 * function has the same parameters as the directive's render() method.
 */
export const directive = (c) => (...values) => ({
    _$litDirective$: c,
    values,
});
/**
 * Base class for creating custom directives. Users should extend this class,
 * implement `render` and/or `update`, and then pass their subclass to
 * `directive`.
 */
export class Directive {
    constructor(partInfo) {
        this._$parent = partInfo._$parent;
        this.__part = partInfo._$part;
        this.__attributeIndex = partInfo._$attributeIndex;
    }
    /** @internal */
    _resolve(props) {
        const { __part, __attributeIndex } = this;
        return resolveDirective(__part, this.update(__part, props), this, __attributeIndex);
    }
    update(_part, props) {
        return this.render(...props);
    }
}
